From b5f011f5016bc34ee16b0b68007d4014e8dc3887 Mon Sep 17 00:00:00 2001
From: qianlongxu <qianlongxu@gmail.com>
Date: Mon, 22 May 2023 18:19:11 +0800
Subject: [PATCH 07] Correct the wrong codecpar->codec_id which read from
 MIME of ID3tags, but the real data was encoded in PNG/JPEG/TIFF

---
 libavformat/id3v2.c   | 7 ++++++-
 libavformat/img2dec.c | 2 +-
 libavformat/mov.c     | 2 +-
 3 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/libavformat/id3v2.c b/libavformat/id3v2.c
index d83716d..98b0e85 100644
--- a/libavformat/id3v2.c
+++ b/libavformat/id3v2.c
@@ -1177,8 +1177,13 @@ int ff_id3v2_parse_apic(AVFormatContext *s, ID3v2ExtraMeta *extra_meta)
         st  = s->streams[s->nb_streams - 1];
         st->codecpar->codec_id   = apic->id;
 
-        if (AV_RB64(st->attached_pic.data) == PNGSIG)
+        if (AV_RB64(st->attached_pic.data) == PNGSIG || AV_RB64(st->attached_pic.data) == MNGSIG) {
             st->codecpar->codec_id = AV_CODEC_ID_PNG;
+        } else if (AV_RB24(st->attached_pic.data) == 0xffd8ff) {
+            st->codecpar->codec_id = AV_CODEC_ID_MJPEG;
+        } else if (AV_RB32(st->attached_pic.data) == 0x49492a00 || AV_RB32(st->attached_pic.data) == 0x4D4D002a) {
+            st->codecpar->codec_id = AV_CODEC_ID_TIFF;
+        }
 
         if (apic->description[0])
             av_dict_set(&st->metadata, "title", apic->description, 0);
diff --git a/libavformat/img2dec.c b/libavformat/img2dec.c
index 2761cb3..a3d8474 100644
--- a/libavformat/img2dec.c
+++ b/libavformat/img2dec.c
@@ -907,7 +907,7 @@ static int png_probe(const AVProbeData *p)
 {
     const uint8_t *b = p->buf;
 
-    if (AV_RB64(b) == 0x89504e470d0a1a0a)
+    if (AV_RB64(b) == 0x89504e470d0a1a0a || AV_RB64(b)  == 0x8a4d4e470d0a1a0a)
         return AVPROBE_SCORE_MAX - 1;
     return 0;
 }
diff --git a/libavformat/mov.c b/libavformat/mov.c
index 606fe22..42eb374 100644
--- a/libavformat/mov.c
+++ b/libavformat/mov.c
@@ -213,7 +213,7 @@ static int mov_read_covr(MOVContext *c, AVIOContext *pb, int type, int len)
     st->priv_data = sc;
 
     if (st->attached_pic.size >= 8 && id != AV_CODEC_ID_BMP) {
-        if (AV_RB64(st->attached_pic.data) == 0x89504e470d0a1a0a) {
+        if (AV_RB64(st->attached_pic.data) == 0x89504e470d0a1a0a || AV_RB64(st->attached_pic.data) == 0x8a4d4e470d0a1a0a) {
             id = AV_CODEC_ID_PNG;
         } else {
             id = AV_CODEC_ID_MJPEG;
-- 
2.39.3 (Apple Git-146)

